### PR TITLE
drivers: phy: add new driver for davicom dm8806 phy

### PR DESCRIPTION
**Pull request by robi251**
New driver for Davicom DM8806 PHY. Only the basic functionality. This is still a draft. please note, that these driver will only working if the related enhancement "Additional PHY API parameter" will be accepted:

### CODE REVIEW

File: drivers/ethernet/phy/Kconfig [hunk 1]
```diff
@@ -40,6 +40,12 @@ config PHY_ADIN2111
 	help
 	  Enable ADIN2111 PHY driver.
 
+config PHY_DM8806
+	bool "DM8806 PHY driver"
+	default y
```
Comment (pdgendt) [inline drivers/ethernet/phy/Kconfig:45]: You need to add `depends on DT_HAS_DAVICOM_DM8806_PHY_ENABLED`

**Comment by pdgendt** [comment]
Also please just force push existing PRs instead of creating new ones.

File: drivers/ethernet/phy/phy_dm8806.c [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT davicom_dm8806_phy
+
+#define LOG_MODULE_NAME eth_dm8806_phy
+#define LOG_LEVEL       CONFIG_ETHERNET_LOG_LEVEL
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
```
Comment (decsny) [inline drivers/ethernet/phy/phy_dm8806.c:14]: you don't need to define macros for the arguments, also this macro has a form that can take both arguments

File: drivers/ethernet/phy/Kconfig (hunk 1 above)
Comment (robi251) [reply drivers/ethernet/phy/Kconfig:45]: Agree - dependency added

File: drivers/ethernet/phy/phy_dm8806.c (hunk 2 above)
Comment (robi251) [reply drivers/ethernet/phy/phy_dm8806.c]: Useless definition are removed

**Comment by pdgendt** [comment]
Squash your commits and force push to this PR, [fixup commits are not allowed](https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow).

**Comment by robi251** [comment]
@pdgendt @decsny HOw do You think about this implementation?

**Comment by pdgendt** [comment]
Please squash all your commits into a single one. It does not make sense to introduce a driver and update it in the same PR.

**Comment by robi251** [comment]
> Please squash all your commits into a single one. It does not make sense to introduce a driver and update it in the same PR.

Of course - I will sqush them to one commit. Give me time until today moon

**Comment by pdgendt** [comment]
> > Please squash all your commits into a single one. It does not make sense to introduce a driver and update it in the same PR.
> 
> Of course - I will sqush them to one commit. Give me time until today moon

Actually you need to split the PHY API change into a separate commit, which might need some more discussion TBH.

**Comment by robi251** [comment]
> > > Please squash all your commits into a single one. It does not make sense to introduce a driver and update it in the same PR.
> > 
> > 
> > Of course - I will sqush them to one commit. Give me time until today moon
> 
> Actually you need to split the PHY API change into a separate commit, which might need some more discussion TBH.

Will it OK, if I will:
1. Remove the completely changes from the PHY API and move it to another one MRQ
2. Squash rest of commits to one commit

**Comment by robi251** [comment]
In separate MRQ:
https://github.com/zephyrproject-rtos/zephyr/pull/79714
we reach the conclusion, that reading MAC address table shouldn't be a part of PHY API, but rather DSI. Because of this reason I deleted it from the driver finaly.

**Comment by decsny** [comment]
> New driver for Davicom DM8806 PHY. Only the basic functionality. This is still a draft. please note, that these driver will only working if the related enhancement "Additional PHY API parameter" will be accepted:

is this still all the case ?

**Comment by robi251** [comment]
> > New driver for Davicom DM8806 PHY. Only the basic functionality. This is still a draft. please note, that these driver will only working if the related enhancement "Additional PHY API parameter" will be accepted:
> 
> is this still all the case ?

Some update is required here:
1. This is now more than basic functionality. Driver is realized full phy API functionality
2. This driver will work only if: [PR81272](https://github.com/zephyrproject-rtos/zephyr/pull/81272) will be approved

Solution was tested on the real hardware with Cortex M4 MCU and is working stable. I don't know if there is more users who can run it on the HW. DM8806 is not so popular :(

**Comment by robi251** [comment]
Hi all.
One simple, but not simple question. I'm trying to figure out why  the pipeline complains because of `phy_dm8806_api`. It says that this API struct is defined but not used as well as the initialization function `phy_dm8806_init`. The construction is the same like in other drivers. What I've missed...?
@tbursztyka or @decsny  - I will be gretefull for any any tips, thank you

**Comment by decsny** [comment]
I didn't look into it but might have something to do with the change to use DEVICE_API macro which was done recently

File: drivers/ethernet/phy/phy_dm8806.c [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+	gpio_pin_interrupt_configure_dt(&cfg->gpio_int, GPIO_INT_EDGE_TO_ACTIVE);
+	if (res != 0) {
+		LOG_WRN("Failed to enable DM8806 interrupt: %i", res);
+		return res;
+	}
+
+	return res;
+}
+
+static const struct ethphy_driver_api phy_dm8806_api = {
```
Comment (pdgendt) [inline drivers/ethernet/phy/phy_dm8806.c:438]:
```suggestion
static DEVICE_API(ethphy, phy_dm8806_api) = {
```

**Review by pdgendt — COMMENTED** [review]
> I didn't look into it but might have something to do with the change to use DEVICE_API macro which was done recently

It shouldn't. That being said, you do need to use the wrapper macro. The warnings you get are if the `FOREACH` macro expands to nothing.

File: drivers/ethernet/phy/Kconfig.dm8806 [hunk 4]
```diff
@@ -0,0 +1,49 @@
+# Copyright 2024 Robert Slawinski <robert.slawinski1@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+# Davicom PHY DM8806 driver configuration options
+
+menuconfig PHY_DM8806
+	bool "Davicom PHY DM8806 driver"
+	default y
+    depends on DT_HAS_DAVICOM_DM8806_PHY_ENABLED
```
Comment (pdgendt) [inline drivers/ethernet/phy/Kconfig.dm8806:9]: Fix indent

File: drivers/ethernet/phy/phy_dm8806.c (hunk 3 above)
Comment (robi251) [reply drivers/ethernet/phy/phy_dm8806.c]: here we go :) Thank you, now it works!

File: drivers/ethernet/phy/Kconfig.dm8806 (hunk 4 above)
Comment (robi251) [reply drivers/ethernet/phy/Kconfig.dm8806]: Fixed!

**Review by pdgendt — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

File: drivers/ethernet/phy/phy_dm8806.c [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+#endif
+	.read = phy_dm8806_reg_read,
+	.write = phy_dm8806_reg_write,
+};
+
+#define DM8806_PHY_DEFINE_CONFIG(n)                                                                \
+	static const struct phy_dm8806_config phy_dm8806_config_##n = {                            \
+		.mdio = DEVICE_DT_GET(DT_INST_BUS(n)),                                             \
+		.phy_addr = DT_INST_REG_ADDR(n),                                                   \
+		.switch_addr = DT_PROP(DT_NODELABEL(dm8806_phy##n), reg_switch),                   \
```
Comment (fabiobaltieri) [inline drivers/ethernet/phy/phy_dm8806.c:452]: this also looks incorrect, it's making an assumption about a nodelabel, I'd imagine this should have been `DT_INST_PROP(n, reg_switch)`

Comment (kartben) [reply drivers/ethernet/phy/phy_dm8806.c:452]: ya it's expecting the device node to have a label for some reason ...

```
dm8806_phy0: ethernet-phy@0 {
			reg = <0>;
			reg-switch = <8>
			...
```

File: drivers/ethernet/phy/phy_dm8806.c [hunk 6]
```diff
@@ -0,0 +1,466 @@
...
+		.gpio_int = GPIO_DT_SPEC_INST_GET(n, interrupt_gpio),                              \
+		.gpio_rst = GPIO_DT_SPEC_INST_GET(n, reset_gpio),                                  \
```
Comment (kartben) [inline drivers/ethernet/phy/phy_dm8806.c:454]: and this is incorrect too...

should be `int_gpios` and `reset_gpios`, plus, the properties are not required so it prob calls for a `GPIO_DT_SPEC_INST_GET_OR` or something ...

**Comment by kartben** [comment]
@jukkar @pdgendt this driver seems very broken. I kind of managed to add it to build_all/ethernet and eventually get it to compile, but I wonder if it should be pulled out or something.

@robi251 Any chance you can contribute whatever improvements/fixes you've certainly had to implement if you're actually using this driver?

**Comment by pdgendt** [comment]
> @jukkar @pdgendt this driver seems very broken. I kind of managed to add it to build_all/ethernet and eventually get it to compile, but I wonder if it should be pulled out or something.
> 
> @robi251 Any chance you can contribute whatever improvements/fixes you've certainly had to implement if you're actually using this driver?

I've opened a PR here https://github.com/zephyrproject-rtos/zephyr/pull/85925

This makes the driver build, no way to test it though. @robi251 can you test with the latest changes?

**Pull request merged**